codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedMapGenerator.kt

```diff
@@ -102,31 +105,57 @@ class UnconstrainedMapGenerator(
                             !valueShape.isDirectlyConstrained(symbolProvider) &&
                             valueShape !is StructureShape &&
                             valueShape !is UnionShape
+                    val constrainedMemberValueSymbol = if (resolvesToNonPublicConstrainedValueType) {
+                        pubCrateConstrainedShapeSymbolProvider.toSymbol(shape.value)
+                    } else {
+                        constrainedShapeSymbolProvider.toSymbol(shape.value)
+                    }
                     val constrainedValueSymbol = if (resolvesToNonPublicConstrainedValueType) {
                         pubCrateConstrainedShapeSymbolProvider.toSymbol(valueShape)
                     } else {
                         constrainedShapeSymbolProvider.toSymbol(valueShape)
                     }
 
                     val constrainedKeySymbol = constrainedShapeSymbolProvider.toSymbol(keyShape)
+                    val epilogueWritable = writable { rust("Ok((k, v))") }
                     val constrainKeyWritable = writable {
                         rustTemplate(
                             "let k: #{ConstrainedKeySymbol} = k.try_into().map_err(Self::Error::Key)?;",
                             "ConstrainedKeySymbol" to constrainedKeySymbol,
                         )
                     }
                     val constrainValueWritable = writable {
-                        rustTemplate(
-                            """
-                            match #{ConstrainedValueSymbol}::try_from(v) {
-                                Ok(v) => Ok((k, v)),
-                                Err(inner_constraint_violation) => Err(Self::Error::Value(k, inner_constraint_violation)),
+                        if (constrainedMemberValueSymbol.isOptional()) {
+                            // The map is `@sparse`.
+                            rustBlock("match v") {
+                                rust("None => Ok((k, None)),")
+                                withBlock("Some(v) =>", ",") {
+                                    // DRYing this up with the else branch below would make this less understandable.
+                                    rustTemplate(
+                                        """
+                                        match #{ConstrainedValueSymbol}::try_from(v) {
+                                            Ok(v) => Ok((k, Some(v))),
+                                            Err(inner_constraint_violation) => Err(Self::Error::Value(k, inner_constraint_violation)),
+                                        }
+                                        """,
+                                        "ConstrainedValueSymbol" to constrainedValueSymbol,
+                                        "Epilogue" to epilogueWritable,
```

Comment:
```suggestion
```

Reply:
This doesn't look to be used here in this block

Reply:
Fixed in https://github.com/awslabs/smithy-rs/pull/2213/commits/b4ae60f85991e647748c2bc1a1bce247761ca1f3.